include/flatbuffers/flatbuffers.h

```diff
@@ -542,57 +547,186 @@ struct String : public Vector<char> {
   }
 };
 
-// Simple indirection for buffer allocation, to allow this to be overridden
-// with custom allocation (see the FlatBufferBuilder constructor).
-class simple_allocator {
+// Allocator interface. This is flatbuffers-specific and meant only for
+// `vector_downward` usage.
+class Allocator {
+ public:
+  virtual ~Allocator() {}
+
+  // Allocate `size` bytes of memory.
+  virtual uint8_t *allocate(size_t size) = 0;
+
+  // Deallocate `size` bytes of memory at `p` allocated by this allocator.
+  virtual void deallocate(uint8_t *p, size_t size) = 0;
+
+  // Reallocate `new_size` bytes of memory, replacing the old region of size
+  // `old_size` at `p`. In contrast to a normal realloc, this grows downwards,
+  // and is intended specifcally for `vector_downward` use.
+  virtual uint8_t *reallocate_downward(uint8_t *old_p,
+                                       size_t old_size,
+                                       size_t new_size) {
+    assert(new_size > old_size);  // vector_downward only grows
+    uint8_t *new_p = allocate(new_size);
+    memcpy(new_p + (new_size - old_size), old_p, old_size);
+    deallocate(old_p, old_size);
+    return new_p;
+  }
+};
+
+// DefaultAllocator uses new/delete to allocate memory regions
+class DefaultAllocator : public Allocator {
+ public:
+  virtual uint8_t *allocate(size_t size) FLATBUFFERS_OVERRIDE {
+    return new uint8_t[size];
+  }
+
+  virtual void deallocate(uint8_t *p, size_t) FLATBUFFERS_OVERRIDE {
+    delete[] p;
+  }
+
+  static DefaultAllocator& instance() {
+    static DefaultAllocator inst;
+    return inst;
+  }
+};
+
+// FlatBuffer is a finished flatbuffer memory region, detached from its
+// builder. The original memory region and allocator are also stored so that
+// the FlatBuffer can manage the memory lifetime.
+class FlatBuffer {
  public:
-  virtual ~simple_allocator() {}
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  FlatBuffer(Allocator *allocator,
+             bool own_allocator,
+             uint8_t *buf,
+             size_t reserved,
+             uint8_t *cur,
+             size_t sz)
+    : allocator_(allocator),
+      own_allocator_(own_allocator),
+      buf_(buf),
+      reserved_(reserved),
+      cur_(cur),
+      size_(sz) {
+    assert(allocator_);
+  }
+
+  FlatBuffer(FlatBuffer &&other)
```

Comment:
What's the appropriate ifdef to use for it? I'm happy to block it out if you want. I can also add a `swap()` method and default ctor so people on older compilers can at least do a manual move it they need to.

I thought it was best to omit the `unique_ptr_t` typedef since it doesn't really behave like a `std::unique_ptr<T>` and might be confusing for newcomers trying to use it as such.